bridge/tracker/tracker.go

```diff
@@ -0,0 +1,390 @@
+	//	get path
+	cwd, err := os.Getwd()
+	if err != nil {
+		return nil, err
+	}
+
+	//	create or load db
+	db, err := leveldb.OpenFile(cwd+"/event_tracker/last_block_number", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return db, nil
+}
+
+//	trackHeader determines the range of block to query
+//	based on the blockNumber of the header and issues an appropriate
+//	eth_getLogs call. If an event of interest was emitted,
+//	it is sent to eventCh.
+func (t *Tracker) trackHeader(header *ethHeader) {
+	//	determine range of blocks to query
+	fromBlock, toBlock, ok := t.calculateRange(header)
+	if !ok {
+		//	we are returning here because the calculated
+		//	range was already queried or the chain is not
+		//	at the desired depth.
+		return
+	}
+
+	t.logger.Info("querying events within block range", "from", fromBlock, "to", toBlock)
+
+	//	fetch logs
+	logs := t.queryEvents(fromBlock, toBlock)
+	t.logger.Info("matched events", "num", len(logs))
+
+	//	notify each matched log event
+	t.notify(logs...)
+}
+
+//	calculateRange determines the next range of blocks
+//	to query for events.
+func (t *Tracker) calculateRange(header *ethHeader) (from, to uint64, ok bool) {
+	//	extract block number from header field
+	//	TODO: polygon-edge header
+	latestHeight := big.NewInt(0).SetUint64(header.Number)
+
+	//	or: TODO ethereum header
+	//latestHeight, _ := types.ParseUint256orHex(&header.Number)
+
+	//	check if block number is at required depth
+	confirmations := big.NewInt(0).SetUint64(t.confirmations)
+	if latestHeight.Cmp(confirmations) < 0 {
+		//	block height less than required
+		t.logger.Debug(
+			"not enough confirmations",
+			"current", latestHeight.Uint64(),
+			"required", confirmations.Uint64())
+
+		return
+	}
+
+	/*	right bound */
+	toBlock := big.NewInt(0).Sub(latestHeight, confirmations)
+
+	/*	left bound */
+	fromBlock := t.loadLastBlock()
+	if fromBlock == nil {
+		fromBlock = toBlock
+	} else {
+		//	increment - last block number was already queried
+		fromBlock = fromBlock.Add(fromBlock, big.NewInt(1))
+	}
+
+	if toBlock.Cmp(fromBlock) < 0 {
+		//	no need to perform any query
+		//	as left bound is higher than the right
+		return
+	}
+
+	return fromBlock.Uint64(), toBlock.Uint64(), true
+}
+
+//	loadLastBlock returns the block number of the last
+// 	block processed by the tracker (if available).
+func (t *Tracker) loadLastBlock() *big.Int {
+	//	read from db
+	bytesLastBlock, err := t.db.Get([]byte(lastQueriedBlockNumber), nil)
+	if err != nil {
+		t.logger.Error("cannot read db", "err", err)
+
+		return nil
+	}
+
+	//	parse
+	lastBlock, ok := big.NewInt(0).SetString(string(bytesLastBlock), 10)
+	if !ok {
+		return nil
+	}
+
+	return lastBlock
+}
+
+//	saveLastBlock stores the number of the last block processed
+//	into the tracker's database.	.
+func (t *Tracker) saveLastBlock(blockNumber *big.Int) error {
+	//	store last processed block number
+	if err := t.db.Put(
+		[]byte(lastQueriedBlockNumber),
```

Comment:
Why not have this be a constant, instead of casting it to `[]byte` every time?

Reply:
[Redefined](https://github.com/0xPolygon/polygon-edge/pull/386/commits/768525864fdabbdbb371131efa3f9f98e7ca80a7)